perfrunbook/system-load-and-compute-headroom.md

```diff
@@ -0,0 +1,120 @@
+# System Load and Compute Headroom
+
+From an economical perspective compute resources should be utilized fully.
+Ideally the system's performance should increase linearly up to 100% CPU load.
+This can be in the context of pure compute or peak load and fail-over resilience.
+Graviton instances and x86-type instances, such as c6i, behave fundamentally different under load.
+Any initial x86 advantage over Graviton starts to disappear beyond 50% system load.
+This effect is caused by symmetric multithreading and the Linux kernel's preferrence of CPU cores over CPU threads.
+In the following this behavior is investigated with a few illustrative workloads.
+
+### Experimental Setup
+
+The test systems are limited to eight cores to limit the maximum packet load needed to saturate them. 
+This is to exclude the influence of the networking infrastructure on the experiment.
+All systems run Amazon Linux 2 with Linux kernel 5.10.118.
+The test system are virtualized instances that use S3 networked storage.
+
+
+#### First experiment
+
+The systems to be tested are an 8 vCPU Intel instance (8375C, 3.5GHz), c6i.2xlarge, and an 8 CPU Graviton 3 instance, c7g.2xlarge.
+OpenSSL source is placed and configured in eight separate directories.
+The test then compiles OpenSSL one to eight times in parallel, limited to one CPU per compile.
+An ideal system would take the exact same time for each compile run.
+
+Results:
+
+The Graviton 3 achieves 96% of the single compile time performance, single CPU baseline when using 100% of the available CPUs.
+The Intel system achieves 63% of the compile time performance of the single compile, single CPU baseline when loaded to 100%.
+
+![](images/system-load/c7g-compared-to-c6i.png)
+
+Note: 
+The average compile time of the c6i instance is 268s, close to the 250s on c7g. 
+The Linux scheduler picks cores over threads, when available, which shows nicely in the steep increase in compile time of the c6i beyond 50% load.
+
+#### Second experiment
+
+The second experiment adds a 64 core Graviton 3, c7g.16xlarge, that serves as a load generator running wrk2.
+Test systems are an 8 CPU c7g.2xlarge Graviton 2 and an 8 vCPU c6i.2xlarge Intel x86/SMT (8375C).
+Here a simple HTTP server, implemented using the Netty/Java framework, is the workload on the systems under test.
+Latency vs actual packets processed is captured alongside the CPU load of the systems, as reported by /proc/stat.
+Of particular interest is the performance at 50% and at maximum sustained load.
+Maximum sustained load is where the response latency is no longer than two times the response time at <20% CPU load.
+This is the reason why the tables below stop at 96% and 94% respectively.
+
+Results:
+
+```
+Graviton 3, c7g.2xlarge, 8 CPUs
+Requests/s, CPU use [%], Latency [us]
+49336, 9, 1142
```

Comment:
If we need the table, can it be a markdown table instead?  Can some of this data be included in the plot somehow as lines+text showing the key points?  i.e. 50%, theoretical 100% load, and actual achieved load?